### PR TITLE
Add minmal SyncManager #56

### LINKED ISSUE #56: SyncManager and SyncDb
## Description

SyncManager synchronizes the client state from the operator.

### Question

How about History Proof?

## References

https://docs.plasma.group/projects/spec/en/latest/src/04-client-architecture/sync-manager.html

### PR DESCRIPTION
**Pull request by syuhei176**
depends on #36 and #77 
Close #56

### CODE REVIEW

File: client/src/sync/sync_manager.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            sync_db: SyncDb::new(T::open(&"sync")),
+        }
+    }
+}
+
+impl<T> SyncManager<T>
+where
+    T: SyncDbTrait,
+{
+    pub fn handle_block_submitted(&self) {
```
Comment (syuhei176) [inline client/src/sync/sync_manager.rs]: `handle_block_submitted` will be called when BlockSubmitted event fired from Commitment Contract.

File: client/src/sync/sync_manager.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+            .is_ok());
+        assert!(sync_manager
+            .remove_deposit_contract(deposit_contract, commit_contract)
+            .is_ok());
+    }
+
+    #[test]
+    fn test_add_and_remove_sync_query() {
+        let sync_manager: SyncManager<SyncDb<CoreDbMemoryImpl>> = Default::default();
+        let deposit_contract: Address = Address::zero();
```
Comment (shogochiai) [inline client/src/sync/sync_manager.rs]: Why not default?

File: client/src/sync/sync_manager.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        let query = StateQuery::new(
+            Address::zero(),
+            predicate_address,
+            Some(0),
+            Some(100),
+            Bytes::new(),
+        );
+
+        assert!(sync_manager
+            .add_sync_query(deposit_contract, &query)
```
Comment (shogochiai) [inline client/src/sync/sync_manager.rs]: 1. StateQuery is for view bulding
2. SyncQuery is StateQuery
3. So, this function is planned to be used in client-side?

Comment (shogochiai) [reply client/src/sync/sync_manager.rs]: Ah, this file is in `/client`, nvm

File: client/src/sync/sync_db.rs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    fn add_sync_query(
+        &self,
+        deposit_contract: Address,
+        state_query: &StateQuery,
+    ) -> Result<(), Error> {
+        let state_query_key: BaseDbKey = state_query.to_hash().into();
+        self.db
+            .root()
+            .bucket(&BaseDbKey::from(&b"sync_queries"[..]))
+            .bucket(&deposit_contract.as_bytes().into())
```
Comment (shogochiai) [inline client/src/sync/sync_db.rs]: For one childchain, one deposit contract isn't it?
Why did you designate to use deposit contract as a bucket?

File: client/src/sync/sync_manager.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            .iter()
+            .fold::<Vec<StateQuery>, _>(vec![], |acc, c| {
+                let mut sync_queries = self.get_sync_queries(*c).ok().unwrap();
+                sync_queries.extend(acc);
+                sync_queries
+            })
+    }
+
+    fn apply_state_query(&self, _state_query: &StateQuery) -> Option<Vec<StateQueryResult>> {
+        Some(vec![])
```
Comment (shogochiai) [inline client/src/sync/sync_manager.rs]: Not inplemented?

File: client/src/sync/sync_db.rs (hunk 4 above)
Comment (syuhei176) [reply client/src/sync/sync_db.rs]: no, there are multiple deposit contracts for each token types.

File: client/src/sync/sync_manager.rs (hunk 5 above)
Comment (syuhei176) [reply client/src/sync/sync_manager.rs]: depends on the https://github.com/cryptoeconomicslab/plasma-rust-framework/issues/9
to get data from operator(aggregator).

File: client/src/sync/sync_manager.rs (hunk 2 above)
Comment (syuhei176) [reply client/src/sync/sync_manager.rs]: I think it's ok now. I will use specific address for test in future.

**Pull request merged**